File: app/src/main/java/me/eycia/msghub_android/MainActivityChannelsAdapter.java
package me.eycia.msghub_android;

import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;

import me.eycia.Notifier;
import me.eycia.api.API;
import me.eycia.api.ChanInfo;

/**
 * Created by eycia on 2/27/16.
 */
public class MainActivityChannelsAdapter extends FragmentPagerAdapter {
    private ChanInfo[] chans;

    private Handler cronHandler = new Handler();
    public Notifier ChansNotifier = new Notifier();

    public void onSaveInstanceState(Bundle savedInstanceState) {
        savedInstanceState.putParcelableArray("chanInfos", chans);
    }

    private void ChangeChansData(final ChanInfo[] cs) {
        chans = cs;
        notifyDataSetChanged();
        ChansNotifier.ChangeData();
    }

    private void CallAPI() {
        new API.ChansInfoTask() {
            @Override
            protected void onSuccess(@NonNull ChanInfo[] chanInfos) {
                ChangeChansData(chanInfos);
            }
        }.execute();
    }

    private Runnable cronUpdateChansInfo = new Runnable() {
        @Override
        public void run() {
            CallAPI();
            cronHandler.postDelayed(this, 10000);
        }
    };

    public MainActivityChannelsAdapter(FragmentManager fm, Bundle savedInstanceState) {
        super(fm);
        chans = new ChanInfo[0];
        cronHandler.post(cronUpdateChansInfo);

        if (savedInstanceState == null) {
            CallAPI();
        } else {
            ChangeChansData((ChanInfo[]) savedInstanceState.getParcelableArray("chanInfos"));
        }
    }

    @Override
    public ChanFragment getItem(int position) {
        ChanFragment cf = ChanFragment.newInstance(chans[position]);
        return cf;
    }

    @Override
    public int getCount() {
        return chans.length;
    }

    @Override
    public CharSequence getPageTitle(int position) {
        return chans[position].Title;
    }
}

File: app/src/main/java/me/eycia/msghub_android/MainActivityChannelsAdapter.kt
package me.eycia.msghub_android

import android.os.Bundle
import android.os.Handler
import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentPagerAdapter
import me.eycia.Notifier
import me.eycia.api.API
import me.eycia.api.ChanInfo

class MainActivityChannelsAdapter(fm: FragmentManager, savedInstanceState: Bundle?) : FragmentPagerAdapter(fm) {
    private var chans: Array<ChanInfo>

    private val cronHandler = Handler()
    var ChansNotifier = Notifier()

    fun onSaveInstanceState(savedInstanceState: Bundle) {
        savedInstanceState.putParcelableArray("chanInfos", chans)
    }

    private fun ChangeChansData(cs: Array<ChanInfo>) {
        chans = cs
        notifyDataSetChanged()
        ChansNotifier.ChangeData()
    }

    private fun CallAPI() {
        object : API.Msgs.ChansInfoTask() {
            override fun onSuccess(result: Array<ChanInfo>) {
                ChangeChansData(result)
            }
        }.execute()
    }

    private val cronUpdateChansInfo = Runnable {
        CallAPI()
        //cronHandler.postDelayed(this, 10000);
    }

    init {
        chans = emptyArray()
        cronHandler.post(cronUpdateChansInfo)

        if (savedInstanceState == null) {
            CallAPI()
        } else {
            ChangeChansData(savedInstanceState.getParcelableArray("chanInfos") as Array<ChanInfo>)
        }
    }

    override fun getItem(position: Int): ChanFragment {
        return ChanFragment.newInstance(chans[position])
    }

    override fun getCount(): Int {
        return chans.size
    }

    override fun getPageTitle(position: Int): CharSequence {
        return chans[position].Title
    }
}
